include/mqtt/endpoint.hpp

```diff
@@ -7426,16 +7422,33 @@ class endpoint : public std::enable_shared_from_this<endpoint<Socket, Mutex, Loc
     template <typename Iterator>
     typename std::enable_if<std::is_convertible<typename Iterator::value_type, char>::value>::type
     restore_serialized_message(packet_id_t /*packet_id*/, Iterator b, Iterator e) {
+        static_assert(
+            std::is_same<
+                typename std::iterator_traits<Iterator>::iterator_category,
+                std::random_access_iterator_tag
+            >::value,
+            "Iterator doesn't support random access"
+        );
+
         if (b == e) return;
 
         auto fixed_header = static_cast<std::uint8_t>(*b);
         switch (get_control_packet_type(fixed_header)) {
         case control_packet_type::publish: {
-            auto sp = std::make_shared<std::string>(b, e);
-            restore_serialized_message(basic_publish_message<PacketIdBytes>(sp->begin(), sp->end()), sp);
+            auto buf = allocate_buffer(b, e);
+            restore_serialized_message(
+                basic_publish_message<PacketIdBytes>(
+                    buf
+                ),
+                buf
+            );
         } break;
         case control_packet_type::pubrel: {
-            restore_serialized_message(basic_pubrel_message<PacketIdBytes>(b, e));
+            restore_serialized_message(
+                basic_pubrel_message<PacketIdBytes>(
+                    buffer(string_view(&*b, static_cast<std::size_t>(std::distance(b, e))))
```

Comment:
A comment explaining why you construct the mqtt::buffer such that it does not own the data would be good to have.

Reply:
I will add the following comment:

```cpp
// basic_pubrel_message have no member variable that type is mqtt::buffer.
// When creating basic_pubrel_message, the constructor just read mqtt::buffer
// and convert to some values.
// So the argument buffer(...) doesn't need to hold the lifetime.
```